### PR TITLE
Feature call destructor for tuples and bug fixes for the destructor process

### PR DESCRIPTION
**Pull request by pmqtt**
This PR includes the following changes:
*  Added destruction process for tuples
*  Fix: In the current version only the last member of a object can be destroyed
*  Fix: In the current version, the destructor of the object is called after each method call 

I hope it is useful

### CODE REVIEW

File: explorer/testdata/destructor/destroy_array.carbon [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+class A{
+    destructor[me: Self]{
+        Print("DESTRUCTOR A {0}",me.n);
+    }
+    var n: i32;
+}
+
+fn Main() -> i32 {
+  var a: [A; 2] = ({.n = 0},{.n = 1});
```
Comment (jonmeow) [inline explorer/testdata/destructor/destroy_array.carbon]: This looks backwards -- I would think the second element of the tuple should be destructed before the first, if we're sticking to typical unrolling of destruction.

File: explorer/testdata/destructor/call_destructor_from_destructor.carbon [hunk 2]
```diff
@@ -30,15 +32,16 @@ class B{
         Print("DESTRUCTOR B {0}",me.n);
     }
     fn Create(x: i32) -> B{
-       return {.n = x, .a = A.Create(3) };
+       return {.n = x, .a1 = A.Create(4),.a2 = A.Create(3) };
```
Comment (jonmeow) [inline explorer/testdata/destructor/call_destructor_from_destructor.carbon:32]: This test changes doesn't look closely related to the rest, am I missing something?

File: explorer/testdata/destructor/dont_call_in_method.carbon [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        Print("TEST");
+    }
+    fn test2[me: Self](){
+        Print("TEST 2");
+    }
+    var n: i32;
+}
+
+fn Main() -> i32 {
+  var a: A = {.n = 1 };
```
Comment (jonmeow) [inline explorer/testdata/destructor/dont_call_in_method.carbon:34]: This isn't doing anything with tuples; is this related to the rest of the changes?

File: explorer/interpreter/action.h [hunk 4]
```diff
@@ -27,11 +27,7 @@ namespace Carbon {
 // not compile-time constants.
 class RuntimeScope {
  public:
-  enum class State {
-    Normal = 0,
-    Destructor = 1,
-    CleanUpped = 2,
-  };
+  enum class State { Normal = 0, Destructor = 1, CleanUpped = 2, Method = 4 };
```
Comment (jonmeow) [inline explorer/interpreter/action.h]: Do these enum values need to have numbers assigned?

File: explorer/interpreter/action.h [hunk 5]
```diff
@@ -80,13 +76,17 @@ class RuntimeScope {
   // Return scope state
   // Normal     = Scope is not bind at the top of a destructor call
   // Destructor = Scope is bind to a destructor call and was not cleaned up,
+  // Method     = Scope is bind at the top of a method
```
Comment (jonmeow) [inline explorer/interpreter/action.h]: Why can't a method be treated as equivalent to other scopes?

**Review by jonmeow — COMMENTED** [review]
I'm a little confused what this is trying to do. The PR description and some code talks about arrays, some about methods. However, is this trying to implement tuple description? (tuples are distinct from arrays)

The difference between tuples and values in a scope seems relatively narrow to me -- it seems to me more like this should be a destructor behavior for tuples, rather than something handled through cleanup through a tuple-specific action.

File: explorer/testdata/destructor/destroy_array.carbon (hunk 1 above)
Comment (pmqtt) [reply explorer/testdata/destructor/destroy_array.carbon]: Sorry, you are right! Done!

File: explorer/testdata/destructor/call_destructor_from_destructor.carbon (hunk 2 above)
Comment (pmqtt) [reply explorer/testdata/destructor/call_destructor_from_destructor.carbon:32]: It is tested that more than one member of the object is destroyed

File: explorer/testdata/destructor/dont_call_in_method.carbon (hunk 3 above)
Comment (pmqtt) [reply explorer/testdata/destructor/dont_call_in_method.carbon:34]: It is tested that the destructor is not called after a method call

File: explorer/interpreter/action.h (hunk 4 above)
Comment (pmqtt) [reply explorer/interpreter/action.h]: Done!

File: explorer/interpreter/action.h (hunk 5 above)
Comment (pmqtt) [reply explorer/interpreter/action.h]: It is iterated through all LValues of a scope. In the case of a method, the me-value is in the scope and therefore the destructor of the object would be called after the method call.

**Comment by jonmeow** [comment]
> The difference between tuples and values in a scope seems relatively narrow to me -- it seems to me more like this should be a destructor behavior for tuples, rather than something handled through cleanup through a tuple-specific action.

@geoffromer for thoughts here. Note, I'm expressing hesitance about the approach of having CleanupAction become this kind of multi-purpose multi-modal thing. But I also think looking for better approaches would take time that I'm not ready to spend on explorer.

**Review by geoffromer — COMMENTED** [review]
Here's my suggestion: we introduce a new kind of `Action` called `DestroyAction`, which is responsible for destroying a single value by invoking the destructor (if any) and then spawning `DestroyAction`s to destroy any sub-values. I'd tentatively prefer for `DestroyAction` to only store pointers to the `Value` being destroyed and the `Declaration` of its type, and then we re-compute things like the class/tuple members on the fly rather than trying to cache them. It probably shouldn't have any mutable state other than `pos`, but I think `pos` should be enough to keep track of where we are in the destruction process.

Does that make sense?

**Comment by pmqtt** [comment]
> Here's my suggestion: we introduce a new kind of `Action` called `DestroyAction`, which is responsible for destroying a single value by invoking the destructor (if any) and then spawning `DestroyAction`s to destroy any sub-values. I'd tentatively prefer for `DestroyAction` to only store pointers to the `Value` being destroyed and the `Declaration` of its type, and then we re-compute things like the class/tuple members on the fly rather than trying to cache them. It probably shouldn't have any mutable state other than `pos`, but I think `pos` should be enough to keep track of where we are in the destruction process.
> 
> Does that make sense?

You mean so (Pseudocode) 
```C
void StepCleanUp(){
    ….
    If(scope.BindToMethod()){
        // Finish
    }
    LValue *lvalue = Heap.Read(cleanUpAction.Locals(pos));
    If(scope.BindToDestructor()){
        //How to call destructor members here ??????????
    }

    Todo_.Spawn(DestroyAction(lValue));
    …
}

void StepDestroy(){
  LValue * lValue = destroyAction.value();
  Swithc(lValue->Kind){
       Case NominalClassValue:
                        //Call Destructor of the object
       Case TupleValue:
                    Return Todo_.Spawn(destroyAction(Tulpe.elements()[pos]));
       Default:
              Break;
   }
   ….
}

```

The idea would also naturally destroy Composite Objects ( Tuples of Tuples), I like that! But I have the problem how do I solve the problem with the members of a class? This can not be done by StepDestroy in case of a NominalClassValue!

**Comment by geoffromer** [comment]
> > Here's my suggestion: we introduce a new kind of `Action` called `DestroyAction`, which is responsible for destroying a single value by invoking the destructor (if any) and then spawning `DestroyAction`s to destroy any sub-values. I'd tentatively prefer for `DestroyAction` to only store pointers to the `Value` being destroyed and the `Declaration` of its type, and then we re-compute things like the class/tuple members on the fly rather than trying to cache them. It probably shouldn't have any mutable state other than `pos`, but I think `pos` should be enough to keep track of where we are in the destruction process.
> > Does that make sense?
> 
> You mean so (Pseudocode)
> 
> ```c
> void StepCleanUp(){
>     ….
>     If(scope.BindToMethod()){
>         // Finish
>     }
>     LValue *lvalue = Heap.Read(cleanUpAction.Locals(pos));
>     If(scope.BindToDestructor()){
>         //How to call destructor members here ??????????
> ```

I don't think `StepCleanUp` should need to give special treatment to destructors here -- a destructor should just behave like a function with a funny name, so cleaning up the destructor scope just means destroying any variables that are local to the destructor itself. With the model I'm suggesting, the process of destroying the members happens after the destructor returns, as subsequent steps of the `DestroyAction`.

I'm realizing that might be very subtly wrong, though: it's possible we want data members to be destroyed before local variables in the destructor, in which case we'd have to do something a little more complicated here. I'm asking on Discord.

> ```c
>     }
> 
>     Todo_.Spawn(DestroyAction(lValue));
>     …
> }
> 
> void StepDestroy(){
>   LValue * lValue = destroyAction.value();
>   Swithc(lValue->Kind){
>        Case NominalClassValue:
>                         //Call Destructor of the object
>        Case TupleValue:
>                     Return Todo_.Spawn(destroyAction(Tulpe.elements()[pos]));
>        Default:
>               Break;
>    }
>    ….
> }
> ```
> 
> The idea would also naturally destroy Composite Objects ( Tuples of Tuples), I like that! But I have the problem how do I solve the problem with the members of a class? This can not be done by StepDestroy in case of a NominalClassValue!

Why not?

File: explorer/interpreter/interpreter.cpp [hunk 6]
```diff
@@ -1934,44 +1944,101 @@ auto Interpreter::StepDeclaration() -> ErrorOr<Success> {
...
+    }
+  }
+  todo_.Pop();
+  return Success();
+}
+
+auto Interpreter::StepCleanUp() -> ErrorOr<Success> {
+  Action& act = todo_.CurrentAction();
+  CleanupAction& cleanup = cast<CleanupAction>(act);
+  if (act.scope()->kind() != RuntimeScope::Kind::Method) {
```
Comment (geoffromer) [inline explorer/interpreter/interpreter.cpp]: It looks like this won't do _any_ cleanup at the top of a method scope, so it skips the `Self` parameter, but also skips all the function parameters, some of which might need to be destroyed. There are also other cases where a local name goes out of scope but the underlying object shouldn't be destroyed, because the name is effectively an alias. So I don't think treating method scopes differently is the way to go here. Instead, we need to do two things:

- Have `CleanupAction` only clean up locals that have _storage_ in the current scope.
- Stop allocating local storage for names like `me` that act as aliases.

To do that, I suggest changing `RuntimeScope::allocations_` to hold the `Value*`s as well as the allocation IDs, so that it has all the information we need to deallocate and destroy each object. Then we have this code loop over `allocations_` instead of `locals_`. To complement that, we can add a `Bind` method to `RuntimeScope`, which takes the same parameters as `Initialize`, but only binds the name to the value, and doesn't allocate storage for it the way `Initialize` does (and doesn't store anything in `allocations_`). Then we can update the logic that handles `me` to use `Bind` instead of `Initialize`.

We'll also need to update a lot of other code to use `Bind` instead of `Initialize`, and we'll probably also need an API for allocating storage without binding a name to it, but I think that can all be a TODO.

Does that make sense?

Comment (pmqtt) [reply explorer/interpreter/interpreter.cpp]: It makes sense and I will try!

File: explorer/interpreter/interpreter.cpp [hunk 7]
```diff
@@ -2093,16 +2097,15 @@ auto Interpreter::StepDestroy() -> ErrorOr<Success> {
...
-        return todo_.Spawn(
-            std::make_unique<DestroyAction>(lvalue, *value, std::nullopt));
-      }
+  if (act.pos() < cleanup.allocations_count()) {
+    auto allocation =
+        act.scope()->allocations()[cleanup.allocations_count() - act.pos() - 1];
+    auto lvalue = arena_->New<LValue>(Address(allocation));
+    SourceLocation source_loc("destructor", 1);
+    auto value = heap_.Read(lvalue->address(), source_loc);
+    if (value.ok()) {
```
Comment (geoffromer) [inline explorer/interpreter/interpreter.cpp:2083]: Should this be `CARBON_RETURN_IF_ERROR`? Or should there be an `else` case? Dropping the error and returning `Success` seems surprising.

File: explorer/interpreter/action.cpp [hunk 8]
```diff
@@ -25,20 +25,18 @@ RuntimeScope::RuntimeScope(RuntimeScope&& other) noexcept
...
   allocations_ = std::exchange(rhs.allocations_, {});
   heap_ = rhs.heap_;
-  destructor_scope_ = rhs.destructor_scope_;
   return *this;
 }
 
 RuntimeScope::~RuntimeScope() {
-  for (AllocationId allocation : allocations_) {
+  for (auto allocation : allocations_) {
     heap_->Deallocate(allocation);
```
Comment (geoffromer) [inline explorer/interpreter/action.cpp:41]:
```suggestion
    // TODO: move this into StepCleanUp
    heap_->Deallocate(allocation);
```

File: explorer/interpreter/action.h [hunk 9]
```diff
@@ -55,6 +48,8 @@ class RuntimeScope {
   void Print(llvm::raw_ostream& out) const;
   LLVM_DUMP_METHOD void Dump() const { Print(llvm::errs()); }
 
+  void Bind(ValueNodeView value_node, Nonnull<const Value*> value);
```
Comment (geoffromer) [inline explorer/interpreter/action.h:52]:
```suggestion
  // Binds `value` as the value of `value_node`.
  void Bind(ValueNodeView value_node, Nonnull<const Value*> value);
```

File: explorer/interpreter/action.h [hunk 10]
```diff
@@ -293,19 +274,48 @@ class DeclarationAction : public Action {
 
 class CleanupAction : public Action {
```
Comment (geoffromer) [inline explorer/interpreter/action.h:278]:
```suggestion
// An Action which implements destroying all local allocations in a scope.
class CleanupAction : public Action {
```

File: explorer/interpreter/action.h [hunk 11]
```diff
@@ -293,19 +274,48 @@ class DeclarationAction : public Action {
...
   static auto classof(const Action* action) -> bool {
     return action->kind() == Kind::CleanUpAction;
   }
 
  private:
-  int locals_count_;
+  int allocations_count_;
+};
+
+class DestroyAction : public Action {
```
Comment (geoffromer) [inline explorer/interpreter/action.h:298]:
```suggestion
// An Action which implements destroying a single value, including all nested values.
class DestroyAction : public Action {
```

File: explorer/interpreter/action.h [hunk 12]
```diff
@@ -293,19 +274,48 @@ class DeclarationAction : public Action {
...
+ public:
+  explicit DestroyAction(Nonnull<const LValue*> lvalue,
```
Comment (geoffromer) [inline explorer/interpreter/action.h:306]: This needs a comment. In particular, it needs to explain the meaning of the arguments, and especially the difference between `value` and `lvalue`

File: explorer/interpreter/heap.cpp [hunk 13]
```diff
@@ -46,6 +46,18 @@ auto Heap::Write(const Address& a, Nonnull<const Value*> v,
   return Success();
 }
 
+auto Heap::GetAllocationId(Nonnull<const Value*> v) const
+    -> std::optional<AllocationId> {
+  auto iter = std::find(values_.begin(), values_.end(), v);
```
Comment (geoffromer) [inline explorer/interpreter/heap.cpp:51]: Hmm. I'm not sure this will work, because there's no guarantee that each entry in `values_` is unique-- the same `Value*` might be stored in multiple locations.

Would it work to instead allow `locals_` to hold any kind of `Value`, not just `LValue`s? I think that would avoid the need for this reverse lookup.

File: explorer/interpreter/interpreter.cpp [hunk 14]
```diff
@@ -796,15 +803,16 @@ auto Interpreter::CallDestructor(Nonnull<const DestructorDeclaration*> fun,
   CARBON_CHECK(method.is_method());
   RuntimeScope method_scope(&heap_);
   BindingMap generic_args;
-  CARBON_CHECK(PatternMatch(&method.me_pattern().value(), receiver,
-                            fun->source_loc(), &method_scope, generic_args,
-                            trace_stream_, this->arena_));
 
+  auto p = &method.me_pattern().value();
```
Comment (geoffromer) [inline explorer/interpreter/interpreter.cpp:804]:
```suggestion
  // TODO: move this logic into PatternMatch, and call it here.
  auto p = &method.me_pattern().value();
```

File: explorer/interpreter/interpreter.cpp [hunk 15]
```diff
@@ -878,9 +886,17 @@ auto Interpreter::CallFunction(const CallExpression& call,
       RuntimeScope method_scope(&heap_);
       BindingMap generic_args;
       // Bind the receiver to the `me` parameter.
-      CARBON_CHECK(PatternMatch(&method.me_pattern().value(), m.receiver(),
-                                call.source_loc(), &method_scope, generic_args,
-                                trace_stream_, this->arena_));
+      auto p = &method.me_pattern().value();
+      if (p->kind() == Value::Kind::BindingPlaceholderValue) {
+        const auto& placeholder = cast<BindingPlaceholderValue>(*p);
```
Comment (geoffromer) [inline explorer/interpreter/interpreter.cpp:889]:
```suggestion
        // TODO: move this logic into PatternMatch
        const auto& placeholder = cast<BindingPlaceholderValue>(*p);
```

File: explorer/interpreter/action.h [hunk 16]
```diff
@@ -55,6 +48,8 @@ class RuntimeScope {
   void Print(llvm::raw_ostream& out) const;
   LLVM_DUMP_METHOD void Dump() const { Print(llvm::errs()); }
 
+  void Bind(ValueNodeView value_node, Nonnull<const Value*> value);
+
   // Allocates storage for `value_node` in `heap`, and initializes it with
   // `value`.
   void Initialize(ValueNodeView value_node, Nonnull<const Value*> value);
```
Comment (geoffromer) [inline explorer/interpreter/action.h:57]:
```suggestion
  // TODO: Update existing callers to use Bind instead, where appropriate.
  void Initialize(ValueNodeView value_node, Nonnull<const Value*> value);
```

File: explorer/interpreter/interpreter.cpp [hunk 17]
```diff
@@ -1995,52 +2011,107 @@ auto Interpreter::StepDeclaration() -> ErrorOr<Success> {
...
+  }
+  if (item->kind() == Value::Kind::TupleValue) {
+    return todo_.Spawn(
+        std::make_unique<DestroyAction>(lvalue, item, std::nullopt));
+  }
+  return std::nullopt;
+}
+
+auto Interpreter::StepDestroy() -> ErrorOr<Success> {
   Action& act = todo_.CurrentAction();
```
Comment (geoffromer) [inline explorer/interpreter/interpreter.cpp:2015]:
```suggestion
  // TODO: find a way to avoid dyn_cast in this code, and instead use static type
  // information the way the compiler would.
  Action& act = todo_.CurrentAction();
```

File: explorer/interpreter/interpreter.cpp [hunk 18]
```diff
@@ -1995,52 +2011,107 @@ auto Interpreter::StepDeclaration() -> ErrorOr<Success> {
...
+              dyn_cast<NominalClassValue>(*destroy_act.value())) {
+        const auto& class_type = cast<NominalClassType>(class_obj->type());
+        const auto& class_dec = class_type.declaration();
+        if (class_dec.destructor().has_value()) {
+          return CallDestructor(*class_dec.destructor(), class_obj);
+        }
+      }
+    }
+  }
+  if (destroy_act.declaration()) {
```
Comment (geoffromer) [inline explorer/interpreter/interpreter.cpp]: Why does the fact that we have a declaration mean that we can return early here?

File: explorer/interpreter/interpreter.cpp [hunk 19]
```diff
@@ -1995,52 +2011,107 @@ auto Interpreter::StepDeclaration() -> ErrorOr<Success> {
...
-          if (class_dec.destructor().has_value()) {
-            return CallDestructor(*class_dec.destructor(), class_obj);
+  DestroyAction& destroy_act = cast<DestroyAction>(act);
+  if (act.pos() == 0) {
+    if (destroy_act.value()) {
+      if (const auto* tuple = dyn_cast<TupleValue>(*destroy_act.value())) {
+        if (tuple->elements().size() > 0) {
+          auto res = DestroyTupleElement(destroy_act.lvalue(), tuple,
+                                         tuple->elements().size() - 1);
+          if (res) {
```
Comment (geoffromer) [inline explorer/interpreter/interpreter.cpp]: Are there situations where `res` is `nullopt`? What's supposed to happen in that case?

File: explorer/interpreter/interpreter.cpp [hunk 20]
```diff
@@ -1995,52 +2011,107 @@ auto Interpreter::StepDeclaration() -> ErrorOr<Success> {
   }
 }
 
-auto Interpreter::StepCleanUp() -> ErrorOr<Success> {
+auto Interpreter::DestroyTupleElement(Nonnull<const LValue*> lvalue,
```
Comment (geoffromer) [inline explorer/interpreter/interpreter.cpp]: I'd prefer to inline the body of this function in the places where it's called, so that it's clear that we're still following the rule that `return` from a "Step" function always calls an `ActionStack` transition method. I think we can do that without duplicating the code -- see my comment below about the `pos() == 0` case.

File: explorer/interpreter/interpreter.cpp [hunk 21]
```diff
@@ -1995,52 +2011,107 @@ auto Interpreter::StepDeclaration() -> ErrorOr<Success> {
...
-      if (act.scope()->DestructionState() < RuntimeScope::State::CleanUpped) {
-        if (const auto* class_obj = dyn_cast<NominalClassValue>(*value)) {
-          const auto& class_type = cast<NominalClassType>(class_obj->type());
-          const auto& class_dec = class_type.declaration();
-          if (class_dec.destructor().has_value()) {
-            return CallDestructor(*class_dec.destructor(), class_obj);
+  DestroyAction& destroy_act = cast<DestroyAction>(act);
+  if (act.pos() == 0) {
+    if (destroy_act.value()) {
+      if (const auto* tuple = dyn_cast<TupleValue>(*destroy_act.value())) {
```
Comment (geoffromer) [inline explorer/interpreter/interpreter.cpp]: I think we should drop this sub-case, and have the `pos() == 0` case do nothing except call the destructor. That way it's a lot clearer why `pos() == 0` is special, and we don't have to factor out `DestroyTupleElement`. That means the first step will be a no-op if `value` isn't a `NominalClassValue`, but I think that's OK.

**Review by geoffromer — COMMENTED** [review]
I think this is getting close. The only thing that really worries me is my comment on `GetAllocationId` -- everything else looks pretty straightforward, and even that one might be OK to resolve with a TODO if necessary.

File: explorer/interpreter/interpreter.cpp (hunk 7 above)
Comment (pmqtt) [reply explorer/interpreter/interpreter.cpp:2083]: I need them to step over uninitialized values

File: explorer/interpreter/interpreter.cpp (hunk 20 above)
Comment (pmqtt) [reply explorer/interpreter/interpreter.cpp:2016]: Impressive that you see such a change in a review!

File: explorer/interpreter/interpreter.cpp (hunk 21 above)
Comment (pmqtt) [reply explorer/interpreter/interpreter.cpp]: Done

Comment (pmqtt) [reply explorer/interpreter/interpreter.cpp]: Impressive that you see such a change in a review! Done!

File: explorer/interpreter/interpreter.cpp (hunk 20 above)
Comment (pmqtt) [reply explorer/interpreter/interpreter.cpp]: DOne!

File: explorer/interpreter/interpreter.cpp (hunk 19 above)
Comment (pmqtt) [reply explorer/interpreter/interpreter.cpp]: Done!

File: explorer/interpreter/interpreter.cpp (hunk 18 above)
Comment (pmqtt) [reply explorer/interpreter/interpreter.cpp]: Removed!

File: explorer/interpreter/action.h (hunk 12 above)
Comment (pmqtt) [reply explorer/interpreter/action.h:306]: Done!

File: explorer/interpreter/heap.cpp (hunk 13 above)
Comment (pmqtt) [reply explorer/interpreter/heap.cpp:51]: First: I have try to change locals from LValue to Value...! But this creates new issues in the method RuntimeScope::Get... To Change this method signature implies that the interpreter must be adapted. 

Second: Currently, the method Bind is called only on the me Value. From this reason I don't think we have a issue... In a later step, we need a solution!

Third: My first implementation of Bind looked so 
```cpp
void RuntimeScope::Bind(ValueNodeView value_node, Nonnull<const Value*> value) {
  CARBON_CHECK(!value_node.constant_value().has_value());
  CARBON_CHECK(value->kind() != Value::Kind::LValue);
    auto id = heap_->AllocateValue(value);
    auto [it, success] =
        locals_.insert({value_node, heap_->arena().New<LValue>(Address(id))});
    CARBON_CHECK(success) << "Duplicate definition of " << value_node.base();
}
```

But I think, this is wrong! I am reight ?

If you have an idea, how to implement Bind. let me know. Currently, I'm really lost!

File: explorer/interpreter/heap_allocation_interface.h [hunk 22]
```diff
@@ -30,6 +30,9 @@ class HeapAllocationInterface {
   // Returns the arena used to allocate the values in this heap.
   virtual auto arena() const -> Arena& = 0;
 
+  virtual auto GetAllocationId(Nonnull<const Value*> v) const
```
Comment (geoffromer) [inline explorer/interpreter/heap_allocation_interface.h:35]:
```suggestion
  // Returns the ID of the first allocation that holds `v`, if one exists.
  // TODO: Find a way to remove this.
  virtual auto GetAllocationId(Nonnull<const Value*> v) const
```

File: explorer/interpreter/heap.cpp (hunk 13 above)
Comment (geoffromer) [reply explorer/interpreter/heap.cpp:51]: OK, let's leave this as a TODO for now. I've added a suggestion for that.

**Review by geoffromer — APPROVED** [review]
Sorry this took so long, but thanks for seeing it through!
